packages/itwinui-react/src/core/utils/hooks/useSyncExternalStore.ts

```diff
@@ -0,0 +1,53 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import * as React from 'react';
+const _React = React; // prevent bundlers from stripping the namespace import
+
+/**
+ * Wrapper around `React.useSyncExternalStore` that uses a shim for React 17.
+ */
+export const useSyncExternalStore =
+  _React.useSyncExternalStore || useSyncExternalStoreShim;
+
+// ----------------------------------------------------------------------------
+
+/**
+ * The shim below is adapted from React's source to make it ESM-compatible.
+ *
+ * Note: This does not use `getServerSnapshot` at all, because there is
+ * apparently no way to check "hydrating" state in pre-18.
+ *
+ * @see https://github.com/facebook/react/tree/main/packages/use-sync-external-store
+ */
+function useSyncExternalStoreShim<T>(
+  subscribe: (onSubscribe: () => void) => () => void,
+  getSnapshot: () => T,
+): T {
+  const value = getSnapshot();
+  const [{ instance }, forceUpdate] = React.useState({
+    instance: { value, getSnapshot },
+  });
```

Comment:
Just wondering, what's the benefit of using `useState({ instance: {…} })` instead of directly using `useState( {…} )`?

Reply:
this is taken directly from the react source code. i'm not exactly sure why they did it this way, but i think it is so they get a new object each time (and thus force a re-render when needed).

Reply:
yup, my understanding is correct. see comment here: https://github.com/facebook/react/blob/2b036d3f1f016fbe8b121d223d96f09e785b97e1/packages/use-sync-external-store/src/useSyncExternalStoreShimClient.js#L82-L84